llvm/projects/ton-compiler/cpp-sdk/tvm/ton_switcher.hpp

```diff
@@ -0,0 +1,279 @@
+
+template<typename T>
+using on_bounced2_t = decltype(&T::_on_bounced);
+template<typename T>
+constexpr bool supports_on_bounced2_v = std::experimental::is_detected_v<on_bounced2_t, T>;
+
+template <class Contract>
+__always_inline Contract load_contract_data() {
+  using DContract = typename Contract::data;
+  Contract c;
+  auto &base = static_cast<DContract &>(c);
+  using data_tup_t = to_std_tuple_t<DContract>;
+  using LinearTup = decltype(make_chain_tuple(data_tup_t{}));
+  parser persist(persistent_data::get());
+  auto data_tuple = parse<LinearTup>(persist);
+  base = to_struct<DContract>(chain_fold_tup<data_tup_t>(data_tuple));
+  return c;
+}
+
+template <class Contract>
+__always_inline void save_contract_data(Contract& contract) {
+  auto &base = static_cast<typename Contract::data &>(contract);
+  auto data_tup = to_std_tuple(base);
+  auto chain_tup = make_chain_tuple(data_tup);
+  persistent_data::set(build(chain_tup).make_cell());
+}
+
+template<bool Internal, class Contract, class IContract, class DContract, unsigned Index, unsigned RestMethods>
+struct dispatch {
+  static const unsigned my_method_id = get_func_id<IContract, Index>();
+
+  __always_inline static void execute(Contract& c, cell msg, slice msg_body, int func_id) {
+    if constexpr (!get_interface_method_getter<IContract, Index>::value) {
+      if (func_id == my_method_id) {
+        using ISmart = typename Contract::base;
+        constexpr auto func = get_interface_method_ptr<Contract, ISmart, Index>::value;
+
+        using Args = get_interface_method_arg_struct<ISmart, Index>;
+        constexpr unsigned args_sz = calc_fields_count<Args>::value;
+
+        parser body_p(msg_body);
+
+        if constexpr (args_sz != 0) {
+          DEBUG_PRINT("parse_args");
+          auto parsed_args = parse<Args>(body_p);
+          DEBUG_PRINT("call_w_args");
+          std::apply_f<func>(std::tuple_cat(std::make_tuple(std::ref(c)),
+                                            to_std_tuple(parsed_args)));
+        } else {
+          DEBUG_PRINT("call_wo_args");
+          (c.*func)();
+        }
+        return;
+      }
+    }
+
+    using next_impl = dispatch<Internal, Contract, IContract, DContract, Index + 1,
+                               RestMethods - 1>;
+    next_impl::execute(c, msg, msg_body, func_id);
+  }
+};
+
+template<bool Internal, class Contract, class IContract, class DContract, unsigned Index>
+struct dispatch<Internal, Contract, IContract, DContract, Index, 0> {
+  __always_inline static void execute(Contract& c, cell msg, slice msg_body, int func_id) {
+    if constexpr (supports_fallback2_v<Contract>) {
+      return c._fallback(msg, msg_body);
+    } else {
+      tvm_throw(error_code::wrong_public_call);
+    }
+  }
+};
+
+template <bool Internal, class Contract, class IContract, class DContract>
+__always_inline int ton_switcher(cell msg, slice msg_body) {
+  bool is_deploy = msg_has_init<Internal>(msg.ctos());
```

Comment:
Just for my curiosity: is the ticktock call a normal call from the user's perspective?

I can see that in the switcher entry point you have special handling of `internal` and `bounced` for example, but not for ticktock. Is this flow a part of `regular_dispatch`?

Reply:
No, ticktock isn't called by users, it is a special transaction, handled by Collator. Ticktock is just a simple method without function id and any user data.
Dispatch, in turn, is a mechanism to find a method by function ID, passed in the inbound message. It is relevant only for ordinary transactions, such as internal or external messages (and getters, but it has own dispatcher).

Reply:
I see, thanks for explanation!